File: interface/src/util/beatport.js
import { formatNumber } from '../validate/track';

/**
 * Beatports Genre mappings are pretty poor. For now, since I purchase *mostly
 * hardcore* remap a few genres automatically.
 */
const genreRemapping = {
  'Hardcore / Hard Techno': 'Hardcore',
  'Hard Dance': 'Hardcore',
};

/**
 * Beatport files contain the beatport ID as the prefix of the file name. We
 * make the assumption here that the ID will always be at least 6 characters
 * long, to help with identification.
 */
const IdRegex = /^[0-9]{5,}(?=_)/;

/**
 * Identify the Beatport ID of a track.
 */
export function identifyId(track) {
  const match = track.filePath.match(IdRegex);

  return match === null ? false : match[0];
}

/**
 * Transform track details to conform to standards when given a beatport
 * details object, including the release and number of tracks.
 */
export function getPartialTrack(origTrack, beatportDetails) {
  const { totalTracks, release } = beatportDetails;
  const track = {};

  track.release = release;

  if (totalTracks === 1) {
    track.album = '';
    track.track = '';
    track.disc  = '';
  }

  if (totalTracks > 1) {
    const trackNumber = origTrack.track.match(/^[0-9]+/);

    // Beatport does not have 'multi-disc' releases as far as I know
    track.disc  = '1/1';
    track.track = formatNumber(`${trackNumber}/${totalTracks}`);
  }

  if (genreRemapping[origTrack.genre] !== undefined) {
    track.genre = genreRemapping[origTrack.genre];
  }

  return track;
}

File: tunes-manager/importer/beatport.py
import bs4
import urllib.request
from urllib.parse import urljoin

import mediafile


# Beatports Genre mappings are pretty poor. For now, since I purchase *mostly
# hardcore* remap a few genres automatically.
genre_remapping = {
    'Hard Dance':             'Hardcore',
    'Hardcore / Hard Techno': 'Hardcore',
}


def has_metadata(media):
    """
    Determine if the provided mediafile contains beatport metadat.
    """
    webpage = media.mg_file.tags.get('WOAF')

    return webpage is not None and 'beatport.com' in webpage.url


def process(media):
    """
    Process the provided media file by querying beatport for additional details
    that beatport does *not* include in the ID3 tags. Returns a dict containing
    the pre-serialized track values which were modified.

    This *modifies* the media file passed in.
    """
    # Beatport seems to prefix the field with this control character
    track_url = media.mg_file.tags.get('WOAF').url.strip('\x03')

    with urllib.request.urlopen(track_url) as response:
        track = bs4.BeautifulSoup(response, 'html.parser')
        release_link = track.select('a.interior-track-release-artwork-link')[0]
        release_url = urljoin('http://www.beatport.com', release_link['href'])

    with urllib.request.urlopen(release_url) as response:
        release = bs4.BeautifulSoup(response, 'html.parser')

    # Get the list of tracks
    tracks = release.select('div.bucket.tracks ul.bucket-items li')

    # Get a mapping of track details (under the artwork)
    details = release.select('ul.interior-release-chart-content-list')[0]
    details = details.select('li')

    details = [i.find_all('span') for i in details]
    details = {s[0].text.lower(): s[1].text.strip() for s in details}

    # Update the track, reporting only modified fields.
    orig = mediafile.serialize(media)

    media.release = details['catalog']

    # Track is a single
    if len(tracks) == 1:
        media.disc  = ''
        media.track = ''
        media.album = ''

    # Track is part of an album (multiple tracks)
    else:
        media.track.total = len(tracks)
        media.disc = '1/1'

    if media.genre in genre_remapping:
        media.genre = genre_remapping[media.genre]

    updated = mediafile.serialize(media)

    # Return only changed values
    return {k: v for k, v in updated.items() if updated[k] != orig[k]}
